flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala

```diff
@@ -0,0 +1,113 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.runtime.aggregate
+
+import java.lang.Iterable
+import java.sql.Timestamp
+
+import org.apache.calcite.runtime.SqlFunctions
+import org.apache.flink.api.common.functions.RichGroupReduceFunction
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow
+import org.apache.flink.types.Row
+import org.apache.flink.util.{Collector, Preconditions}
+
+/**
+  * It is used for sliding windows on batch for count-windows. It takes a prepared input row,
+  * pre-aggregates (pre-tumbles) rows, aligns the window start, and replicates or omits records
+  * for different panes of a sliding window.
+  *
+  * @param aggregates aggregate functions
+  * @param groupingKeysLength number of grouping keys
+  * @param preTumblingSize number of records to be aggregated (tumbled) before emission
+  * @param windowSize window size of the sliding window
+  * @param windowSlide window slide of the sliding window
+  * @param returnType return type of this function
+  */
+class DataSetSlideCountWindowAggReduceGroupFunction(
+    private val aggregates: Array[Aggregate[_]],
+    private val groupingKeysLength: Int,
+    private val preTumblingSize: Long,
+    private val windowSize: Long,
+    private val windowSlide: Long,
+    @transient private val returnType: TypeInformation[Row])
+  extends RichGroupReduceFunction[Row, Row]
+  with ResultTypeQueryable[Row] {
+
+  private var output: Row = _
+  private var outWindowStartIndex: Int = _
+
+  override def open(config: Configuration) {
+    Preconditions.checkNotNull(aggregates)
+    // add one field to store window start count
+    val partialRowLength = groupingKeysLength +
+      aggregates.map(_.intermediateDataType.length).sum + 1
+    output = new Row(partialRowLength)
+    outWindowStartIndex = partialRowLength - 1
+  }
+
+  override def reduce(records: Iterable[Row], out: Collector[Row]): Unit = {
+    var count: Long = 0
+
+    val iterator = records.iterator()
+
+    while (iterator.hasNext) {
+      val record = iterator.next()
+      // reset aggregates after completed tumbling
+      if (count % preTumblingSize == 0) {
+        // initiate intermediate aggregate value.
+        aggregates.foreach(_.initiate(output))
+      }
+
+      // merge intermediate aggregate value to buffer.
+      aggregates.foreach(_.merge(record, output))
+
+      count += 1
+
+      // trigger tumbling evaluation
+      if (count % preTumblingSize == 0) {
```

Comment:
I thought about this again. I think it is ok not to emit incomplete tumbling windows. If the count is not reached in the tumbling window it does not trigger a sliding window.